### PR TITLE
Fix validation for MinValue and MaxValue attributes

### LINKED ISSUE #234: [Bug]: Verifying the minvalue or maxvalue attributes always fails when using floating point params
### Description

When setting up a slash command with a floating point parameter (say a float, double, or decimal) and you add a MinValue or MaxValue attribute, building the command tree fails with the error message `A non-numerical parameter may not specify a minimum or maximum value.`. The problem is with CommandTreeExtensions, line 611.

### Steps to Reproduce

- Create a slash command with a float or double parameter with a MinValue or MaxValue attribute.
- Attempt to check slash commands support and/or update commands
- Observe that the mentioned error message is thrown unless the attributes are removed.

### Expected Behavior

The error should not be thrown, allowing the command tree to be built. (Barring any other issues with the registered commands)

### Current Behavior

The error message mentioned in the description is returned: `A non-numerical parameter may not specify a minimum or maximum value.`

### Library / Runtime Information

.NET 6.0.8
Remora.Discord 2022.48

### PR DESCRIPTION
**Pull request by murl-digital**
closes #234 

Added some tests and fixed the bug I found. While working on this I also noticed that the min and max value attributes can't be applied to integers, should this be changed?

### CODE REVIEW

File: Remora.Discord.Commands/Extensions/CommandTreeExtensions.cs [hunk 1]
```diff
@@ -608,7 +608,7 @@ ILocalizationProvider localizationProvider
             var minValue = parameter.Parameter.GetCustomAttribute<MinValueAttribute>();
             var maxValue = parameter.Parameter.GetCustomAttribute<MaxValueAttribute>();
 
-            if (discordType is not Integer or Number && (minValue is not null || maxValue is not null))
+            if ((discordType is not Integer && discordType is not Number) && (minValue is not null || maxValue is not null))
```
Comment (MazeXP) [inline Remora.Discord.Commands/Extensions/CommandTreeExtensions.cs:611]:
```suggestion
            if ((discordType is not (Integer or Number) && (minValue is not null || maxValue is not null))
```

File: Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs [hunk 2]
```diff
@@ -730,6 +730,144 @@ public void CreatesMinAndMaxLengthConstrainedParametersCorrectly()
...
+
+                var result = tree.CreateApplicationCommands();
+                ResultAssert.Successful(result);
+
+                var commands = result.Entity;
+
+                var command = commands.Single();
+                var parameter = command.Options.Value.Single();
+
+                Assert.Equal(1l, parameter.MaxValue.Value);
```
Comment (MazeXP) [inline Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs:776]:
```suggestion
                Assert.Equal(1L, parameter.MaxValue.Value);
```

File: Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs [hunk 3]
```diff
@@ -730,6 +730,144 @@ public void CreatesMinAndMaxLengthConstrainedParametersCorrectly()
...
+
+                var result = tree.CreateApplicationCommands();
+                ResultAssert.Successful(result);
+
+                var commands = result.Entity;
+
+                var command = commands.Single();
+                var parameter = command.Options.Value.Single();
+
+                Assert.Equal(0l, parameter.MinValue.Value);
```
Comment (MazeXP) [inline Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs:753]:
```suggestion
                Assert.Equal(0L, parameter.MinValue.Value);
```

File: Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs [hunk 4]
```diff
@@ -730,6 +730,144 @@ public void CreatesMinAndMaxLengthConstrainedParametersCorrectly()
...
```
Comment (MazeXP) [inline Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs:799]:
```suggestion
                Assert.Equal(0L, parameter.MinValue.Value);
```

File: Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs [hunk 5]
```diff
@@ -730,6 +730,144 @@ public void CreatesMinAndMaxLengthConstrainedParametersCorrectly()
...
+                Assert.Equal(1l, parameter.MaxValue.Value);
```
Comment (MazeXP) [inline Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs:800]:
```suggestion
                Assert.Equal(1L, parameter.MaxValue.Value);
```

File: Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs [hunk 6]
```diff
@@ -730,6 +730,144 @@ public void CreatesMinAndMaxLengthConstrainedParametersCorrectly()
...
+
+                var result = tree.CreateApplicationCommands();
+                ResultAssert.Successful(result);
+
+                var commands = result.Entity;
+
+                var command = commands.Single();
+                var parameter = command.Options.Value.Single();
+
+                Assert.Equal(0l, parameter.MinValue.Value);
```
Comment (MazeXP) [inline Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs:822]:
```suggestion
                Assert.Equal(0L, parameter.MinValue.Value);
```

File: Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs [hunk 7]
```diff
@@ -730,6 +730,144 @@ public void CreatesMinAndMaxLengthConstrainedParametersCorrectly()
...
+
+                var result = tree.CreateApplicationCommands();
+                ResultAssert.Successful(result);
+
+                var commands = result.Entity;
+
+                var command = commands.Single();
+                var parameter = command.Options.Value.Single();
+
+                Assert.Equal(1l, parameter.MaxValue.Value);
```
Comment (MazeXP) [inline Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs:845]:
```suggestion
                Assert.Equal(1L, parameter.MaxValue.Value);
```

File: Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs [hunk 8]
```diff
@@ -730,6 +730,144 @@ public void CreatesMinAndMaxLengthConstrainedParametersCorrectly()
...
+
+                var result = tree.CreateApplicationCommands();
+                ResultAssert.Successful(result);
+
+                var commands = result.Entity;
+
+                var command = commands.Single();
+                var parameter = command.Options.Value.Single();
+
+                Assert.Equal(0l, parameter.MinValue.Value);
```
Comment (MazeXP) [inline Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs:868]:
```suggestion
                Assert.Equal(0L, parameter.MinValue.Value);
```

File: Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs [hunk 9]
```diff
@@ -730,6 +730,144 @@ public void CreatesMinAndMaxLengthConstrainedParametersCorrectly()
...
+                Assert.Equal(1l, parameter.MaxValue.Value);
```
Comment (MazeXP) [inline Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs:869]:
```suggestion
                Assert.Equal(1L, parameter.MaxValue.Value);
```

**Review by MazeXP — CHANGES_REQUESTED** [review]
LGTM, just fix those minor things

**Comment by MazeXP** [comment]
Can you elaborate more on "min and max value attributes can't be applied to integers" please

**Comment by murl-digital** [comment]
> Can you elaborate more on "min and max value attributes can't be applied to integers" please

I could just be completely missing something, but it seems like the attributes only take ulongs, longs, floats, or doubles as parameters. It took me until right about now that that probably doesn't matter since that's seperate to how parameters are parsed for commands.

**Comment by MazeXP** [comment]
Technically it does support integers as well, as its value will be implicitly casted to long or ulong.
Same would also count for float and double.

Maybe @Nihlus could elaborate why the attribute uses the types in the OneOf.
Related line: https://github.com/Remora/Remora.Discord/blob/main/Remora.Discord.Commands/Attributes/MinValueAttribute.cs#L39

File: Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs (hunk 9 above)
Comment (murl-digital) [reply Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs]: addressed in [f6692a0](https://github.com/Remora/Remora.Discord/pull/235/commits/f6692a0b8af281b38ee0d9e2509409f58165c9a9)

File: Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs (hunk 8 above)
Comment (murl-digital) [reply Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs]: addressed in [f6692a0](https://github.com/Remora/Remora.Discord/pull/235/commits/f6692a0b8af281b38ee0d9e2509409f58165c9a9)

File: Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs (hunk 7 above)
Comment (murl-digital) [reply Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs]: addressed in [f6692a0](https://github.com/Remora/Remora.Discord/pull/235/commits/f6692a0b8af281b38ee0d9e2509409f58165c9a9)

File: Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs (hunk 6 above)
Comment (murl-digital) [reply Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs]: addressed in [f6692a0](https://github.com/Remora/Remora.Discord/pull/235/commits/f6692a0b8af281b38ee0d9e2509409f58165c9a9)

File: Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs (hunk 5 above)
Comment (murl-digital) [reply Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs]: addressed in [f6692a0](https://github.com/Remora/Remora.Discord/pull/235/commits/f6692a0b8af281b38ee0d9e2509409f58165c9a9)

File: Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs (hunk 4 above)
Comment (murl-digital) [reply Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs]: addressed in [f6692a0](https://github.com/Remora/Remora.Discord/pull/235/commits/f6692a0b8af281b38ee0d9e2509409f58165c9a9)

File: Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs (hunk 3 above)
Comment (murl-digital) [reply Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs:753]: addressed in [f6692a0](https://github.com/Remora/Remora.Discord/pull/235/commits/f6692a0b8af281b38ee0d9e2509409f58165c9a9)

File: Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs (hunk 2 above)
Comment (murl-digital) [reply Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs]: addressed in [f6692a0](https://github.com/Remora/Remora.Discord/pull/235/commits/f6692a0b8af281b38ee0d9e2509409f58165c9a9)

File: Remora.Discord.Commands/Extensions/CommandTreeExtensions.cs (hunk 1 above)
Comment (murl-digital) [reply Remora.Discord.Commands/Extensions/CommandTreeExtensions.cs]: addressed in [bf6bc7d](https://github.com/Remora/Remora.Discord/pull/235/commits/bf6bc7d7ca4252876a7e0f3d1081ba7dcf646b8a)

File: Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs (hunk 3 above)
Comment (MazeXP) [inline Tests/Remora.Discord.Commands.Tests/Extensions/CommandTreeExtensionTests.cs:753]:
```suggestion
                Assert.Equal(0L, parameter.MinValue.Value);
```

**Review by MazeXP — CHANGES_REQUESTED** [review]

**Pull request merged**